File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int n=s.length();
        int maxLen=0,l=0,maxFreq=0;
        vector<int> hash(26,0);
        for(int r=0;r<n;r++)
        {
            int ind=s[r]-'A';
            hash[ind]++;
            maxFreq=max(maxFreq,hash[ind]);
            if(r-l+1-maxFreq>k)
            {
                int ind=s[l]-'A';
                hash[ind]--;
                l++;
            }
            if(r-l+1-maxFreq<=k) maxLen=max(maxLen,r-l+1);
        }
        return maxLen;
    }
};

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int n = s.length();
        int l = 0, maxFreq = 0, maxLength = 0;
        int[] hash = new int[26];

        for(int r = 0;r < n; r++) {
            int index = s.charAt(r) - 'A';
            hash[index]++;
            maxFreq = Math.max(maxFreq,hash[index]);
            int length = r - l + 1;

            if(length - maxFreq > k) {
                hash[s.charAt(l) - 'A']--;
                l++;
            }

            if(length - maxFreq <= k) {
                maxLength = Math.max(maxLength,length);
            }
        }
        return maxLength;
    }
}